test/test-users.js

```
@@ -0,0 +1,402 @@
+            res.body.length.should.equal(2); //there are 2 dummy users in our DB
+
+            res.body[0].should.be.an('object');
+            res.body[0].should.have.property('username');
+            res.body[0].should.have.property('firstName');
+            res.body[0].should.have.property('lastName');
+
+            // no password returned
+            res.body[0].should.not.have.property('password');
+          });
+      });
+    });
+
+    describe('POST', () => {
+      it('should allow adding a user', () => {
+        let newUser = {
+          username: 'new.user@example.com',
+          password: 'password',
+          firstName: 'new',
+          lastName: 'user',
+        };
+        return chai.request(app).post('/users').send(newUser)
+          .then(res => {
+            // Check that an empty object is returned
+            res.should.have.status(201);
+            res.type.should.equal('application/json');
+            res.charset.should.equal('utf-8');
+
+            res.should.have.header('location');
+
+            res.body.should.be.an('object');
+            res.body.should.be.empty;
+
+            let params = pattern["/users/:userId"].match(res.headers.location);
+            // Fetch the user from the database, using the location header to get the ID
+            return User.findById(params.userId).exec();
+          })
+          .then(result => {
+            // Check that the user exists in the database
+            should.exist(result);
+            result.should.have.property('username');
+            result.username.should.equal(newUser.username);
+          });
+      });
+      it('should reject users without a username', () => {
+        let badUser = {
+          password: 'password',
+          firstName: 'bad',
+          lastName: 'user',
+        };
+        let spy = chai.spy();
+        // Add a user without a username
+        return chai.request(app).post('/users').send(badUser)
+          .then(spy)
```

Why do we need a spy here?

---

We're verifying the spy wasn't called. I post request without a username should return an error so the `.then(spy)` shouldn't be called